src/store/reducers.ts

```diff
@@ -8,50 +11,60 @@ import {
   DELETE_HABIT,
   UPDATE_STREAK,
   UPDATE_CURRENT_USER,
-  REGISTER_USER
-} from "../actions";
+  REGISTER_USER,
+  State,
+  ActionTypes,
+  UserState,
+} from "./types";
 
-const initialState = {
+const initialState: State = {
   initialized: false, // false means the state hasn't been read from localstorage yet
   currentUser: null, // Null means no one is logged in, a string identifies a username
-  userStates: {}
+  userStates: {},
 };
 
 const STATE_KEY = "state";
 
 // Write redux state to local storage
-export const writeStateMiddleware = store => next => action => {
+export const writeStateMiddleware: Middleware<{}, State> = (store) => (
+  next
+) => (action: ActionTypes) => {
   let result = next(action);
   const { initialized, ...state } = store.getState();
   window.localStorage.setItem(STATE_KEY, JSON.stringify(state));
   return result;
 };
 
 // Retrieve Redux state from local storage
-export const readState = () => {
+export const readState = (): State => {
   try {
     const stateJson = window.localStorage.getItem(STATE_KEY);
-    const parsedState = JSON.parse(stateJson);
-    return decodeState(parsedState);
-  } catch {
+    if (stateJson) {
+      // TODO: verify that stateJSON parsing succeeded and is of type StateJSON
+      const parsedState: StateJSON = JSON.parse(stateJson);
+      return decodeState(parsedState);
+    }
+  } catch (error) {
     /* eslint-disable-next-line no-console */
-    console.log("Failed to read state from local storage");
+    console.log("Failed to read state from local storage", error);
```

Comment:
This could use the more informative `console.error`:

```suggestion
    console.error("Failed to read state from local storage", error);
```